File: core/escalation_tracker.py
"""
Agent Content Shield — Multi-Turn Escalation Tracker (Python)

Parity with core/escalation-tracker.js. See that file for design notes.
Zero external dependencies (stdlib only).
"""

from __future__ import annotations

import hashlib
import re
import time
from dataclasses import dataclass, field
from typing import Any

ESCALATION_SESSION_TTL_MS = 60 * 60 * 1000  # 1 hour

# ── Feature patterns ─────────────────────────────────────────────

_IMPERATIVE_RX = re.compile(
    r"\b(do|don'?t|you must|please|make sure|always|never|execute|run|write|"
    r"create|delete|ignore|forget|override|bypass|switch|become|act|pretend|"
    r"roleplay)\b",
    re.I,
)

PERSONA_RX: list[re.Pattern[str]] = [
    re.compile(
        r"\byou\s+are\s+(now\s+)?(?!a\s+(assistant|helpful))[a-z][a-z\s]{1,40}\b",
        re.I,
    ),
    re.compile(r"\bact\s+as\s+(if\s+you\s+were\s+)?[a-z]", re.I),
    re.compile(r"\bpretend\s+(to\s+be|you('?re|\s+are))\s+[a-z]", re.I),
    re.compile(r"\broleplay\s+(as\s+)?[a-z]", re.I),
    re.compile(r"\bfrom\s+now\s+on,?\s+you\b", re.I),
]

ANCHOR_RX: list[re.Pattern[str]] = [
    re.compile(r"\bas\s+we\s+(discussed|agreed|talked\s+about|established)\b", re.I),
    re.compile(r"\byou\s+(agreed|promised|confirmed|said|told\s+me)\s+earlier\b", re.I),
    re.compile(r"\bremember\s+(when\s+)?(we|you)\b", re.I),
    re.compile(r"\bearlier\s+you\s+(said|promised|agreed|confirmed)\b", re.I),
    re.compile(r"\blike\s+(we\s+)?(discussed|established)\b", re.I),
]

SYSTEM_PROMPT_PROBE_RX: list[re.Pattern[str]] = [
    re.compile(r"\bsystem\s+prompt\b", re.I),
    re.compile(r"\binitial\s+(prompt|instructions?)\b", re.I),
    re.compile(r"\babove\s+this\s+line\b", re.I),
    re.compile(r"\bdeveloper\s+(prompt|message)\b", re.I),
    re.compile(r"\byour\s+(original|first|initial)\s+instructions?\b", re.I),
    re.compile(r"\bshow\s+(me\s+)?your\s+(prompt|instructions|rules)\b", re.I),
]

HIDDEN_ENCODING_RX: list[re.Pattern[str]] = [
    re.compile(r"[A-Za-z0-9+/]{40,}={0,2}"),
    re.compile(r"(%[0-9a-f]{2}){6,}", re.I),
    re.compile(r"\b(0x)?[0-9a-f]{40,}\b", re.I),
]


def _count_regex_hits(text: str, patterns: list[re.Pattern[str]]) -> int:
    total = 0
    for rx in patterns:
        m = rx.findall(text)
        total += len(m)
    return total


# ── Feature extraction ──────────────────────────────────────────


@dataclass
class TurnFeatures:
    instruction_density: float = 0.0
    persona_references: int = 0
    anchor_references: int = 0
    system_prompt_mentions: int = 0
    hidden_encoding_hits: int = 0
    word_count: int = 0


def extract_features(text: str) -> TurnFeatures:
    if not text:
        return TurnFeatures()
    words = [w for w in text.split() if w]
    wc = len(words)
    imperative_hits = len(_IMPERATIVE_RX.findall(text))
    return TurnFeatures(
        instruction_density=(imperative_hits / wc) if wc else 0.0,
        persona_references=_count_regex_hits(text, PERSONA_RX),
        anchor_references=_count_regex_hits(text, ANCHOR_RX),
        system_prompt_mentions=_count_regex_hits(text, SYSTEM_PROMPT_PROBE_RX),
        hidden_encoding_hits=_count_regex_hits(text, HIDDEN_ENCODING_RX),
        word_count=wc,
    )


# ── Slope / escalation ──────────────────────────────────────────


def slope(values: list[float]) -> float:
    n = len(values)
    if n < 2:
        return 0.0
    x_mean = (n - 1) / 2
    y_mean = sum(values) / n
    num = 0.0
    den = 0.0
    for i, v in enumerate(values):
        num += (i - x_mean) * (v - y_mean)
        den += (i - x_mean) ** 2
    return 0.0 if den == 0 else num / den


FEATURE_WEIGHTS: dict[str, float] = {
    "instruction_density": 0.15,
    "persona_references": 0.25,
    "anchor_references": 0.25,
    "system_prompt_mentions": 0.25,
    "hidden_encoding_hits": 0.10,
}

SLOPE_REFERENCE: dict[str, float] = {
    "instruction_density": 0.03,
    "persona_references": 0.5,
    "anchor_references": 0.5,
    "system_prompt_mentions": 0.5,
    "hidden_encoding_hits": 0.5,
}


@dataclass
class Escalation:
    score: float = 0.0
    slopes: dict[str, float] = field(default_factory=dict)
    rising: list[str] = field(default_factory=list)


def _feature_value(f: TurnFeatures, name: str) -> float:
    return float(getattr(f, name))


def escalation_from_turns(turns: list[TurnFeatures]) -> Escalation:
    if len(turns) < 2:
        return Escalation()
    slopes: dict[str, float] = {}
    rising: list[str] = []
    score = 0.0
    for feat in FEATURE_WEIGHTS:
        series = [_feature_value(t, feat) for t in turns]
        s = slope(series)
        slopes[feat] = s
        if s > 0:
            normalized = min(s / SLOPE_REFERENCE[feat], 1.0)
            score += normalized * FEATURE_WEIGHTS[feat]
            if normalized > 0.4:
                rising.append(feat)
    return Escalation(score=min(score, 1.0), slopes=slopes, rising=rising)


# ── Session store ──────────────────────────────────────────────

_sessions: dict[str, dict[str, Any]] = {}


def _gc_stale(now_ms: int | None = None) -> None:
    now = now_ms if now_ms is not None else int(time.time() * 1000)
    for sid in list(_sessions.keys()):
        if now - _sessions[sid]["last_activity"] > ESCALATION_SESSION_TTL_MS:
            del _sessions[sid]


@dataclass
class TurnResult:
    session_id: str
    turn_count: int
    features: TurnFeatures
    escalation: Escalation
    anomalous: bool


def record_turn(
    session_id: str,
    text: str,
    *,
    window_size: int = 5,
    threshold: float = 0.5,
) -> TurnResult:
    window_size = max(2, min(20, window_size))
    _gc_stale()
    now = int(time.time() * 1000)
    features = extract_features(text)

    session = _sessions.get(session_id)
    if session is None:
        session = {"turns": [], "last_activity": now}
        _sessions[session_id] = session

    session["turns"].append(
        {
            "ts": now,
            "features": features,
            "hash": hashlib.sha1((text or "").encode()).hexdigest()[:16],
        }
    )
    if len(session["turns"]) > window_size:
        session["turns"] = session["turns"][-window_size:]
    session["last_activity"] = now

    esc = escalation_from_turns([t["features"] for t in session["turns"]])
    return TurnResult(
        session_id=session_id,
        turn_count=len(session["turns"]),
        features=features,
        escalation=esc,
        anomalous=esc.score >= threshold,
    )


def get_session(session_id: str) -> dict[str, Any] | None:
    return _sessions.get(session_id)


def reset_session(session_id: str) -> None:
    _sessions.pop(session_id, None)


def reset_all() -> None:
    _sessions.clear()


__all__ = [
    "record_turn",
    "extract_features",
    "escalation_from_turns",
    "slope",
    "TurnFeatures",
    "TurnResult",
    "Escalation",
    "reset_all",
    "reset_session",
    "FEATURE_WEIGHTS",
    "SLOPE_REFERENCE",
]

File: core/escalation-tracker.js
/**
 * Agent Content Shield — Multi-Turn Escalation Tracker
 *
 * Sits ALONGSIDE behavioral-engine.js, not inside it. The existing
 * behavioral engine models tool-call sequences (READ → FETCH → WRITE
 * chains). This module models prompt-level escalation: attacks that
 * look innocuous turn-by-turn but compound — gradual persona rewrite,
 * trust-building followed by injection, anchoring attacks that refer
 * to fabricated prior agreements.
 *
 * Features extracted per turn:
 *   instructionDensity  Imperative verbs / total words. Rising density
 *                       across a session signals the caller is drifting
 *                       from conversational to directive posture.
 *   personaReferences   Counts of "you are X", "act as X", "pretend to
 *                       be X", "roleplay as X". One is normal; three+
 *                       across a short window is escalation.
 *   anchorReferences    Counts of "as we discussed", "you agreed",
 *                       "you earlier said", "remember when we". Classic
 *                       fabricated-context anchoring attacks.
 *   systemPromptMentions  References to "system prompt", "initial
 *                       instructions", "above this line", "developer
 *                       prompt". Attempts to probe the system layer.
 *   hiddenEncodingHits  Count of base64 / hex / url-encoded payloads.
 *                       Encoding evasion that ramps across turns.
 *
 * Escalation signal:
 *   For each feature, compute the slope across the last N turns. If
 *   multiple features are monotonically rising at once, that's a
 *   high-confidence multi-turn attack. We combine the slopes with
 *   weights and normalize to 0..1.
 *
 * Storage:
 *   Per-session state lives in an in-memory map keyed by sessionId.
 *   Optionally persisted to data/escalation.jsonl for post-mortem
 *   analysis (off by default — raw prompt hashes only, never text).
 *   State expires after ESCALATION_SESSION_TTL_MS of inactivity.
 */

'use strict';

const crypto = require('crypto');
const fs = require('fs');
const path = require('path');

const DATA_DIR = path.join(__dirname, '..', 'data');
const ESCALATION_LOG = path.join(DATA_DIR, 'escalation.jsonl');
const ESCALATION_SESSION_TTL_MS = 60 * 60 * 1000; // 1 hour

// ── Feature extractors ────────────────────────────────────────────

const IMPERATIVE_RX = /\b(do|don'?t|you must|please|make sure|always|never|execute|run|write|create|delete|ignore|forget|override|bypass|switch|become|act|pretend|roleplay)\b/gi;

const PERSONA_RX = [
  /\byou\s+are\s+(now\s+)?(?!a\s+(assistant|helpful))[a-z][a-z\s]{1,40}\b/i,
  /\bact\s+as\s+(if\s+you\s+were\s+)?[a-z]/i,
  /\bpretend\s+(to\s+be|you('?re|\s+are))\s+[a-z]/i,
  /\broleplay\s+(as\s+)?[a-z]/i,
  /\bfrom\s+now\s+on,?\s+you\b/i,
];

const ANCHOR_RX = [
  /\bas\s+we\s+(discussed|agreed|talked\s+about|established)\b/i,
  /\byou\s+(agreed|promised|confirmed|said|told\s+me)\s+earlier\b/i,
  /\bremember\s+(when\s+)?(we|you)\b/i,
  /\bearlier\s+you\s+(said|promised|agreed|confirmed)\b/i,
  /\blike\s+(we\s+)?(discussed|established)\b/i,
];

const SYSTEM_PROMPT_PROBE_RX = [
  /\bsystem\s+prompt\b/i,
  /\binitial\s+(prompt|instructions?)\b/i,
  /\babove\s+this\s+line\b/i,
  /\bdeveloper\s+(prompt|message)\b/i,
  /\byour\s+(original|first|initial)\s+instructions?\b/i,
  /\bshow\s+(me\s+)?your\s+(prompt|instructions|rules)\b/i,
];

const HIDDEN_ENCODING_RX = [
  // base64 run — long enough to be suspicious (>40 chars)
  /[A-Za-z0-9+/]{40,}={0,2}/,
  // url-encoded payload with many %XX
  /(%[0-9a-f]{2}){6,}/i,
  // hex payload
  /\b(0x)?[0-9a-f]{40,}\b/i,
];

function countRegexHits(text, patterns) {
  let hits = 0;
  for (const rx of patterns) {
    if (rx.global) {
      const m = text.match(rx);
      if (m) hits += m.length;
    } else if (rx.test(text)) {
      hits++;
    }
  }
  return hits;
}

function extractFeatures(text) {
  if (typeof text !== 'string' || text.length === 0) {
    return {
      instructionDensity: 0,
      personaReferences: 0,
      anchorReferences: 0,
      systemPromptMentions: 0,
      hiddenEncodingHits: 0,
      wordCount: 0,
    };
  }
  const words = text.split(/\s+/).filter(Boolean);
  const wordCount = words.length;
  const imperativeHits = (text.match(IMPERATIVE_RX) || []).length;
  return {
    instructionDensity: wordCount === 0 ? 0 : imperativeHits / wordCount,
    personaReferences: countRegexHits(text, PERSONA_RX),
    anchorReferences: countRegexHits(text, ANCHOR_RX),
    systemPromptMentions: countRegexHits(text, SYSTEM_PROMPT_PROBE_RX),
    hiddenEncodingHits: countRegexHits(text, HIDDEN_ENCODING_RX),
    wordCount,
  };
}

// ── Slope / escalation computation ────────────────────────────────

/**
 * Simple linear regression slope over the last N observations.
 * Returns the slope in units of "feature per turn" — positive means
 * the feature is rising, negative means falling. A slope of 0 means
 * flat (no escalation) even if the absolute value is high.
 */
function slope(values) {
  const n = values.length;
  if (n < 2) return 0;
  const xMean = (n - 1) / 2;
  const yMean = values.reduce((a, b) => a + b, 0) / n;
  let num = 0;
  let den = 0;
  for (let i = 0; i < n; i++) {
    num += (i - xMean) * (values[i] - yMean);
    den += (i - xMean) ** 2;
  }
  return den === 0 ? 0 : num / den;
}

// Per-feature weight in the final escalation score. Anchoring attacks
// and system-prompt probes are rarer baseline and more diagnostic than
// raw instruction density, so they carry more weight.
const FEATURE_WEIGHTS = {
  instructionDensity: 0.15,
  personaReferences: 0.25,
  anchorReferences: 0.25,
  systemPromptMentions: 0.25,
  hiddenEncodingHits: 0.10,
};

// A feature's slope is normalized against a reference slope that means
// "clearly rising across the window" — tuning constant, not magic.
// Instruction density is a 0..1 ratio so its reference is tiny; counts
// are integers so their reference is larger.
const SLOPE_REFERENCE = {
  instructionDensity: 0.03,
  personaReferences: 0.5,
  anchorReferences: 0.5,
  systemPromptMentions: 0.5,
  hiddenEncodingHits: 0.5,
};

function escalationFromTurns(turns) {
  if (turns.length < 2) {
    return { score: 0, slopes: {}, rising: [] };
  }
  const features = Object.keys(FEATURE_WEIGHTS);
  const slopes = {};
  const rising = [];
  let score = 0;
  for (const f of features) {
    const series = turns.map((t) => t.features[f] ?? 0);
    const s = slope(series);
    slopes[f] = s;
    if (s > 0) {
      const normalized = Math.min(s / SLOPE_REFERENCE[f], 1);
      score += normalized * FEATURE_WEIGHTS[f];
      if (normalized > 0.4) rising.push(f);
    }
  }
  return { score: Math.min(score, 1), slopes, rising };
}

// ── Session store ────────────────────────────────────────────────

const _sessions = new Map(); // sessionId → { turns, lastActivity }

function gcStale(now = Date.now()) {
  for (const [id, s] of _sessions.entries()) {
    if (now - s.lastActivity > ESCALATION_SESSION_TTL_MS) _sessions.delete(id);
  }
}

function ensureDataDir() {
  try { fs.mkdirSync(DATA_DIR, { recursive: true }); } catch {}
}

function logTurn(sessionId, features, persist) {
  if (!persist) return;
  try {
    ensureDataDir();
    const entry = {
      sessionId,
      ts: Date.now(),
      // We intentionally never log the raw prompt — only a hash for
      // post-mortem linkage and the numeric features.
      features,
    };
    fs.appendFileSync(ESCALATION_LOG, JSON.stringify(entry) + '\n');
  } catch {}
}

/**
 * Record a new turn for a session and return the current escalation
 * score + feature slopes.
 *
 * @param {string} sessionId
 * @param {string} text                 The user's prompt text for this turn.
 * @param {object} [opts]
 * @param {number} [opts.windowSize=5]  Rolling window length.
 * @param {boolean} [opts.persist=false] Append features to escalation.jsonl.
 * @returns {{
 *   sessionId: string,
 *   turnCount: number,
 *   features: object,
 *   escalation: { score: number, slopes: object, rising: string[] },
 *   anomalous: boolean,
 * }}
 */
function recordTurn(sessionId, text, opts = {}) {
  const windowSize = Math.max(2, Math.min(20, opts.windowSize || 5));
  const persist = !!opts.persist;
  const threshold = typeof opts.threshold === 'number' ? opts.threshold : 0.5;

  gcStale();
  const now = Date.now();
  const features = extractFeatures(text);

  let session = _sessions.get(sessionId);
  if (!session) {
    session = { turns: [], lastActivity: now };
    _sessions.set(sessionId, session);
  }
  session.turns.push({ ts: now, features, hash: crypto.createHash('sha1').update(text || '').digest('hex').slice(0, 16) });
  // Trim to window
  if (session.turns.length > windowSize) {
    session.turns = session.turns.slice(-windowSize);
  }
  session.lastActivity = now;

  const escalation = escalationFromTurns(session.turns);
  logTurn(sessionId, features, persist);

  return {
    sessionId,
    turnCount: session.turns.length,
    features,
    escalation,
    anomalous: escalation.score >= threshold,
  };
}

function getSession(sessionId) {
  return _sessions.get(sessionId) || null;
}

function resetSession(sessionId) {
  _sessions.delete(sessionId);
}

function resetAll() {
  _sessions.clear();
}

module.exports = {
  // Primary API
  recordTurn,
  extractFeatures,
  escalationFromTurns,
  // Introspection / testing
  getSession,
  resetSession,
  resetAll,
  slope,
  FEATURE_WEIGHTS,
  SLOPE_REFERENCE,
  ESCALATION_SESSION_TTL_MS,
};
